lib/parlour/type_parser.rb

```diff
@@ -0,0 +1,609 @@
+        current = T.unsafe(start)
+        indeces.each do |index|
+          raise IndexError, 'path does not exist' if index >= current.to_a.length
+          current = current.to_a[index]
+        end
+        current
+      end
+    end
+
+    extend T::Sig
+
+    sig { params(ast: Parser::AST::Node, unknown_node_errors: T::Boolean).void }
+    # Creates a new {TypeParser} from whitequark/parser AST.
+    #
+    # @param [Parser::AST::Node] The AST.
+    # @param [Boolean] unknown_node_errors Whether to raise an error if a node
+    #   of an unknown kind is encountered. If false, the node is simply ignored;
+    #   if true, a parse error is raised. Setting this to true is likely to
+    #   raise errors for lots of non-RBI Ruby code, but setting it to false
+    #   could miss genuine typed objects if Parlour or your code contains a bug.
+    def initialize(ast, unknown_node_errors: false)
+      @ast = ast
+      @unknown_node_errors = unknown_node_errors
+    end
+
+    sig { params(filename: String, source: String).returns(TypeParser) }
+    # Creates a new {TypeParser} from a source file and its filename.
+    #
+    # @param [String] filename A filename. This does not need to be an actual
+    #   file; it merely identifies this source.
+    # @param [String] source The Ruby source code.
+    # @return [TypeParser]
+    def self.from_source(filename, source)
+      buffer = Parser::Source::Buffer.new(filename)
+      buffer.source = source
+      
+      TypeParser.new(Parser::CurrentRuby.new.parse(buffer))
+    end
+
+    sig { returns(Parser::AST::Node) }
+    # @return [Parser::AST::Node] The AST which this type parser should use.
+    attr_accessor :ast
+
+    sig { returns(T::Boolean) }
+    # @return [Boolean] Whether to raise an error if a node of an unknown kind
+    #   is encountered. 
+    attr_reader :unknown_node_errors
+
+    # Parses the entire source file and returns the resulting root namespace.
+    #
+    # @return [RbiGenerator::Namespace] The root namespace of the parsed source.
+    sig { returns(RbiGenerator::Namespace) }
+    def parse_all
+      root = RbiGenerator::Namespace.new(DetachedRbiGenerator.new)
+      root.children.concat(parse_path_to_object(NodePath.new([])))
+      root
+    end
+
+    # Given a path to a node in the AST, parses the object definitions it
+    # represents and returns it, recursing to any child namespaces and parsing
+    # any methods within.
+    #
+    # If the node directly represents several nodes, such as being a 
+    # (begin ...) node, they are all returned.
+    #
+    # @param [NodePath] path The path to the namespace definition. Do not pass
+    #   any of the other parameters to this method in an external call.
+    # @return [Array<RbiGenerator::RbiObject>] The objects the node at the path
+    #    represents, parsed into an RBI generator object.
+    sig { params(path: NodePath, is_within_eigen: T::Boolean).returns(T::Array[RbiGenerator::RbiObject]) }
```

Comment:
What's eigen?

Reply:
I should probably rename this to `is_within_eigen_class` to be more clear.
The eigenclass is the Ruby object you're accessing when you do `class << self` from within a class.